daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala

```diff
@@ -34,83 +34,77 @@ import scala.reflect.ClassTag
 final class ImmArray[+A] private (
     private val start: Int,
     val length: Int,
-    array: mutable.ArraySeq[A]
+    private val array: Array[AnyRef]
 ) {
   self =>
 
-  def iterator: Iterator[A] = {
-    var cursor = start
+  def iterator: Iterator[A] = new Iterator[A] {
+    private[this] var i = 0
 
-    new Iterator[A] {
-      override def hasNext: Boolean = cursor < start + self.length
+    override def hasNext: Boolean = i < self.length
 
-      override def next(): A = {
-        val x = array(cursor)
-        cursor += 1
-        x
-      }
+    override def next(): A = {
+      val x = uncheckedGet(i)
+      i += 1
+      x
     }
   }
 
-  def reverseIterator: Iterator[A] = {
-    var cursor = start + self.length - 1
+  def reverseIterator: Iterator[A] = new Iterator[A] {
+    private[this] var i = start + self.length - 1
 
-    new Iterator[A] {
-      override def hasNext: Boolean = cursor >= start
+    override def hasNext: Boolean = i >= start
 
-      override def next(): A = {
-        val x = array(cursor)
-        cursor -= 1
-        x
-      }
+    override def next(): A = {
+      val x = array(i).asInstanceOf[A]
+      i -= 1
+      x
     }
   }
 
   /** O(1), crashes on out of bounds */
   def apply(idx: Int): A =
-    if (idx >= length) {
+    if (idx >= length)
       throw new IndexOutOfBoundsException("index out of bounds in ImmArray apply")
-    } else {
+    else
       uncheckedGet(idx)
-    }
 
-  private def uncheckedGet(idx: Int): A = array(start + idx)
+  private def uncheckedGet(idx: Int): A = array(start + idx).asInstanceOf[A]
+
+  private[data] def copyToAnyRefArray[B >: A](
+      dst: Array[AnyRef],
+      dstStart: Int,
+      dstLen: Int,
+  ): Unit =
+    System.arraycopy(array, start, dst, dstStart, length min dstLen)
 
   /** O(n) */
-  def copyToArray[B >: A](dst: Array[B], dstStart: Int, dstLen: Int): Unit = {
-    for (i <- 0 until Math.max(length, dstLen)) {
-      dst(dstStart + i) = uncheckedGet(i)
-    }
-  }
+  def copyToArray[B >: A](dst: Array[B], dstStart: Int, dstLen: Int): Unit =
+    Array.copy(array, start, dst, dstStart, length min dstLen)
 
   /** O(n) */
-  def copyToArray[B >: A](xs: Array[B]): Unit = {
+  def copyToArray[B >: A](xs: Array[B]): Unit =
     copyToArray(xs, 0, length)
-  }
 
   /** O(1), crashes on empty list */
-  def head: A = this(0)
+  def head: A = apply(0)
 
   /** O(1), crashes on empty list */
-  def last: A = this(length - 1)
+  def last: A = apply(length - 1)
 
   /** O(1), crashes on empty list */
-  def tail: ImmArray[A] = {
-    if (length < 1) {
+  def tail: ImmArray[A] =
+    if (isEmpty)
       throw new RuntimeException("tail on empty ImmArray")
-    } else {
+    else
       new ImmArray(start + 1, length - 1, array)
```

Comment:
We might end up with an `ImmAray` with `length == 0` here. Do we want that? It would needlessly hold a pointer to the underlying buffer.

Reply:
fixed.